Tutorial: writing application definitions, building readers (#744)
FAIRmat-NFDI/pynxtools: docs/how-tos/nexus/write-an-application-definition.md

# Write an application definition

An application definition is a formal contract specifying which fields and groups a NeXus file must, should, or may contain for a given experimental technique. Write one when no existing definition covers your measurement — consult the [NeXus definitions catalogue](https://fairmat-nfdi.github.io/nexus_definitions/) first.

For this how-to guide, we assume that you are familiar with:
- the NeXus data model — see [Learn > NeXus > A primer on NeXus](../../learn/nexus/nexus-primer.md)
- naming rules in NeXus — see [Learn > NeXus > # Rules for storing data in NeXus](../../learn/nexus/nexus-rules.md)

For a step-by-step walkthrough of the same material see the tutorial at [Tutorials > Writing your first application definition](../../tutorial/writing-an-application-definition.md).

---

## Step 1 — Design the data contract

> Identify what information is needed for *standard measurements and analysis* in your field — nothing more, nothing less. Assign each piece an optionality level:

Reviewer: ```suggestion
Identify which concepts are needed for *standard measurements and analysis* in your field — nothing more, nothing less. Assign each piece a presence constraint:
```

                       recommended="true">
                    <doc>Horizontal pixel positions relative to the detector centre.</doc>
                    <dimensions rank="1"><dim index="1" value="n_x"/></dimensions>
                </field>
                <field name="y_pixel_offset" type="NX_FLOAT" units="NX_LENGTH"
                       recommended="true">
                    <doc>Vertical pixel positions relative to the detector centre.</doc>
                    <dimensions rank="1"><dim index="1" value="n_y"/></dimensions>
                </field>
            </group>

        </group>

        <group name="interference_pattern" type="NXdata"> <!-- (8) -->
            <doc>Default plot: the recorded 2D interference pattern.</doc>
            <attribute name="signal" value="data"/>
            <attribute name="axes" value="y_pixel_offset x_pixel_offset"/>
            <field name="data" type="NX_NUMBER" units="NX_ANY">
                <dimensions rank="2">
                    <dim index="1" value="n_y"/>
                    <dim index="2" value="n_x"/>
                </dimensions>
            </field>

>             <field name="x_pixel_offset" type="NX_FLOAT" units="NX_LENGTH">

Author: I do not understand this comment, there's a unit here?

   modern convention.
3. **NeXus types** — always specify `type="NX_DATE_TIME"`, `type="NX_FLOAT"`, etc. The default is
   `NX_CHAR`.
4. **Unit categories** — use `NX_WAVELENGTH`, `NX_LENGTH`, `NX_ENERGY`, etc., *not* raw strings
   like `"m"`. Full list in the
   [NeXus units reference](https://manual.nexusformat.org/nxdl-types.html#unit-categories).
5. **Optionality** — no attribute = required; `recommended="true"` = soft requirement;
   `optional="true"` = free.
6. **Instance vs. concept name** — `name="double_slit"` is the *instance* name written to the HDF5
   file; `type="NXslit"` is the *concept*. See [naming rules](../../learn/nexus/nexus-rules.md).
7. **`<dimensions>`** — rank and `dim` values must be consistent. Use symbolic names from
   `<symbols>`.
8. **`NXdata`** — always include a default-plot group with `@signal` and `@axes` attributes. In a
   real HDF5 file these fields would typically be hard links into the detector group.

### Alternative: write in YAML with nyaml

The [`nyaml`](https://github.com/FAIRmat-NFDI/nyaml) tool (developed by FAIRmat) lets you write the same definition in YAML and convert it to NXDL XML:

```bash

> pip install nyaml

Reviewer: ok
